### PR TITLE
Publish sfContactId on sendConfirmationEmail

### PR DESCRIPTION
**Pull request by QuarpT**
Part of ET to Braze CMT migration.
We are going to store sfContactIds in Braze.
Required by membship-workflow. 
see https://github.com/guardian/membership-workflow/pull/143

### CODE REVIEW

File: handlers/digital-subscription-expiry/src/test/scala/com/gu/digitalSubscriptionExpiry/DigitalSubscriptionExpiryStepsTest.scala [hunk 1]
```diff
@@ -232,22 +232,19 @@ class DeserialiserTest extends FlatSpec with Matchers {
...
     val json = """{"apiToken": "a", "apiClientId": "b"}"""
-    val actualRequest = Json.parse(json).validate[UrlParams]
 
     Json.parse(json).validate[UrlParams] should be(JsSuccess(UrlParams(false)))
 
   }
 
   it should "manage with the noActivation param being false" in {
     val json = """{"apiToken": "a", "apiClientId": "b", "noActivation": "false"}"""
-    val actualRequest = Json.parse(json).validate[ApiGatewayRequest]
```
Comment (johnduffell) [inline handlers/digital-subscription-expiry/src/test/scala/com/gu/digitalSubscriptionExpiry/DigitalSubscriptionExpiryStepsTest.scala]: that was weird, good spot

File: handlers/new-product-api/src/main/scala/com/gu/newproduct/api/addsubscription/Handler.scala [hunk 2]
```diff
@@ -110,7 +111,7 @@ object Steps {
     getCustomerData: ZuoraAccountId => ApiGatewayOp[ContributionCustomerData],
     contributionValidations: (ValidatableFields, Currency) => ValidationResult[AmountMinorUnits],
     createSubscription: ZuoraCreateSubRequest => ClientFailableOp[SubscriptionName],
-    sendConfirmationEmail: ContributionsEmailData => AsyncApiGatewayOp[Unit]
+    sendConfirmationEmail: (Option[SfContactId], ContributionsEmailData) => AsyncApiGatewayOp[Unit]
```
[the remark this answers is not in the document]
Comment (johnduffell) [reply handlers/new-product-api/src/main/scala/com/gu/newproduct/api/addsubscription/Handler.scala]: This is very useful feedback, as the idea of the codebase is so that new people can maintain the code without too much trouble so your opinion is more valuable than ours!

I'm not sure about the abstraction is required, some is needed for sure, but this is a response to the previous systems where we had a massive "touchpoint backend" object in each project that literally knew about every system that we could possibly talk to, and grew sideways over time. eg
https://github.com/guardian/subscriptions-frontend/blob/master/app/services/TouchpointBackends.scala#L35
 every bit of code was directly coupled to all the layers below it, so it was hard to change things from the bottom up.

When we started with lambdas, we couldn't use that code because it was inseparable and large but we still wanted sharing so we looked for a more mix and match approach.  In the lambdas we are trying to make each bit of code only know about exactly what it needed and at only work on one level of abstraction, and take in relatively minimal data structures.  So this does seem to make writing unit tests very clear and easy as a side benefit.

Unfortunately we have seemed to end up with a lot of wiring code to connect together all the different parts in the right way to work (as we have to inject dependencies within dependencies to go through every layer of abstraction), and as you mention we end up with lots of higher order functions which prove to be hard to follow.  It ends up looking like a version of haskell with poorer syntax.

We have been experimenting with single abstract method pattern (ie. a trait with one method which scala converts automatically) and that lets us give it a name. i.e.
https://github.com/guardian/zuora-auto-cancel/blob/788ac37ab660d360fa1ac5e8b05b2039514ae76c/handlers/sf-contact-merge/src/main/scala/com/gu/sf_contact_merge/Handler.scala#L75-L78

We are very keen that things are easy to understand for newcomers to the project which is hard when only two of us are the main contributors.  So it would be great to get some specific suggestions from you, even if it's just go through and use traits everywhere a function is passed.

any other thoughts @pvighi ?

File: handlers/new-product-api/src/main/scala/com/gu/newproduct/api/addsubscription/email/ETPayload.scala [hunk 3]
```diff
@@ -20,10 +20,11 @@ object ETPayload {
 
   implicit def writes[A: Writes] = Json.writes[ETPayload[A]]
 
-  def apply[A](email: String, fields: A, name: DataExtensionName): ETPayload[A] =
+  def apply[A](email: String, fields: A, name: DataExtensionName, sfContactId: Option[String]): ETPayload[A] =
     ETPayload(
       To = CTo(email, email, CContactAttributes(fields)),
-      DataExtensionName = name.value
+      DataExtensionName = name.value,
+      SfContactId = sfContactId
```
Comment (pvighi) [inline handlers/new-product-api/src/main/scala/com/gu/newproduct/api/addsubscription/email/ETPayload.scala]: Is the sfContactId needed at this level and in the fields like all other personal data ?

**Comment by johnduffell** [comment]
looks great code wise, same comments about the PR description apply as in https://github.com/guardian/membership-workflow/pull/143 but I think I get the gist just not the detail which would be very useful
Thanks for the comments on readability too, I have replied above
Patricio has been working on that area of code so address his comment but from my point of view once he is happy it's a 👍 from me

File: handlers/new-product-api/src/main/scala/com/gu/newproduct/api/addsubscription/email/ETPayload.scala (hunk 3 above)
Comment (paulbrown1982) [reply handlers/new-product-api/src/main/scala/com/gu/newproduct/api/addsubscription/email/ETPayload.scala]: I feel the Salesforce Contact ID here is essentially a handle for the attribute information (i.e. metadata), and is needed by the Members Data API to augment the metadata even more in order to send to Braze.  I feel it's leaky abstraction for the Membership Workflow service to pluck it out of the [CContactAttributes] data map.

File: handlers/new-product-api/src/main/scala/com/gu/newproduct/api/addsubscription/Handler.scala (hunk 2 above)
[the remark this answers is not in the document]
Comment (QuarpT) [reply handlers/new-product-api/src/main/scala/com/gu/newproduct/api/addsubscription/Handler.scala]: `single abstract method pattern` would help 👍 
It's the same as wrapping primitives in AnyVal case classes, which the project currently does 😄

[the remark this answers is not in the document]
Comment (QuarpT) [reply handlers/new-product-api/src/main/scala/com/gu/newproduct/api/addsubscription/Handler.scala]: I think the `TouchpointBackend` is less complicated than the `operationForEffects` for comprehension.

`every bit of code was directly coupled to all the layers below it, so it was hard to change things from the bottom up.`
I don't think abstracting every function from the application makes things easier to change.  A change at the bottom still propagates to the top. The types are still very tightly coupled.

[the remark this answers is not in the document]
Comment (johnduffell) [reply handlers/new-product-api/src/main/scala/com/gu/newproduct/api/addsubscription/Handler.scala]: first comment, very insightful, yes it's basically giving a name in the type system to something more generic and it is helpful to give high level meaning.  Based on your feedback we will move over to that in all new or touched code assuming @pvighi agrees

regarding the second comment, you are right about changes propagating through all layers explicitly which is a pain.  Having thought about what it means I think it depends on your definition of complex.  We discussed a little yesterday about how an http client could be considered more powerful than a function from SFContactId => LastName (say `GetName`).  Therefore any function that takes an http client is more complex than one that takes a `GetName` due to the fact that the http client one could actually go off and grab anything from the internet.
If we later decide that the function that takes an http client should also use an internet based surname formatter we wouldn't need to change the parameters (so the change would be easier), however in the latter case we would have to build and pass in a LastName=>FormattedLastName function.  That means we have restricted what the function can do by only passing in things that it will really use.
In my mind, the TouchPointBackend is an object that already has all of those functions created and grouped together in some way.  But we pass them all in regardless of whether they are all actually used.

Perhaps we could have a more middle ground where touchpoint backend isn't really a global object but some kind of loose grouping of functions needed that we can filter down at will.  I'll carry on thinking about this topic, as your comments are helpful!

[the remark this answers is not in the document]
Comment (pvighi) [reply handlers/new-product-api/src/main/scala/com/gu/newproduct/api/addsubscription/Handler.scala]: Yeah I've been noticing that the pattern we have been using gets a little complicated with bigger lambdas. 
I still think having things just depend on what they actually need makes them simpler and easier to test but the wiring code gets out of control when you try to actually put everything together. I agree we should find some grouping that makes sense and makes it easier to understand

[the remark this answers is not in the document]
Comment (johnduffell) [reply handlers/new-product-api/src/main/scala/com/gu/newproduct/api/addsubscription/Handler.scala]: just wondering what's the difference between touchpointbackend and a bootstrap object?  is the TouchpointBackend just a bloated bootstrap object?  And likewise for the result of OperationForEffects?  Is it just a bootstrap object that uses some specific bespoke code to build in each different case.

TouchpointBackend gives excellent sharing of all the building up code, as there's only one set and it does the same every time, but I agree that some parts of operationForEffects could be subdivided and shared with some more thought.

File: handlers/new-product-api/src/main/scala/com/gu/newproduct/api/addsubscription/Handler.scala [hunk 4]
```diff
@@ -148,7 +149,7 @@ object Steps {
       contacts = customerData.contacts,
       paymentMethod = customerData.paymentMethod
     )
-    _ <- sendConfirmationEmail(voucherEmailData).recoverAndLog("send voucher confirmation email")
+    _ <- sendConfirmationEmail(customerData.account.sfContactId, voucherEmailData).recoverAndLog("send voucher confirmation email")
```
Comment (pvighi) [inline handlers/new-product-api/src/main/scala/com/gu/newproduct/api/addsubscription/Handler.scala]: sfContactId here could also have been part of VoucherEmailData (and the same with contributions)

File: handlers/new-product-api/src/main/scala/com/gu/newproduct/api/addsubscription/Handler.scala (hunk 2 above)
[the remark this answers is not in the document]
Comment (QuarpT) [reply handlers/new-product-api/src/main/scala/com/gu/newproduct/api/addsubscription/Handler.scala]: You mentioned the problem with TouchpointBackend was that it grew too large and was being shared everywhere. I was just thinking of alternatives. 

Thanks for the detailed explanations, they've been very insightful @johnduffell 😄

[the remark this answers is not in the document]
Comment (johnduffell) [reply handlers/new-product-api/src/main/scala/com/gu/newproduct/api/addsubscription/Handler.scala]: yep thanks for your comments too we will certainly be able to improve how we write this project based on your feedback 🥇

File: handlers/new-product-api/src/test/scala/com/gu/newproduct/api/addsubscription/zuora/GetAccountEffectsTest.scala [hunk 5]
```diff
@@ -21,6 +21,7 @@ class GetAccountEffectsTest extends FlatSpec with Matchers {
     } yield res
     val expected = Account(
       identityId = Some(IdentityId("30000549")),
+      sfContactId = Some(SfContactId("sfContactId")),
```
Comment (johnduffell) [inline handlers/new-product-api/src/test/scala/com/gu/newproduct/api/addsubscription/zuora/GetAccountEffectsTest.scala]: @QuarpT for future reference, these test have to be run manually by doing `sbt effectsTest:test` after getting membership aws credentials.  Just gives extra confidence that the api calls are all working.

They use example accounts in dev zuora/SF/identity so they need to have the appropriate value.  We have fixed so no issues or anything, just for reference in future.

Of course these tests should run automatically but we have to get credentials to teamcity so it can access stuff.
